### ISSUE TITLE
ENH: Add ModuleInfoLoader ... to support automatic registration of entity classes (via generated code using querybean generator)

### ISSUE THREAD

**Issue by rbygrave** [ebean-orm/ebean#1885]
## Off by default
Initially by default this is turned OFF (via `ServerConfig.loadModuleInfo`)

## About
Applications using querybean generator will have generated code (an implementation of `ModuleInfoLoader`) that will be able to register entity classes.

This will be nice to have turned on such that we then don't need to write that code ourselves (to register entity bean classes) or use classpath scanning. 

An example of the generated ModuleInfoLoader is:

```java
package org.example.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.ebean.config.ModuleInfo;
import io.ebean.config.ModuleInfoLoader;

@ModuleInfo(entities={"org.example.domain.Address","org.example.domain.Contact","org.example.domain.Content","org.example.domain.Country","org.example.domain.Customer","org.example.domain.DMachine","org.example.domain.DMachineStats","org.example.domain.DMachineStatsAgg","org.example.domain.Order","org.example.domain.OrderDetail","org.example.domain.OrderShipment","org.example.domain.OrderTotals","org.example.domain.Product"})
public class _ebean$ModuleInfo implements ModuleInfoLoader {

  @Override
  public List<Class<?>> entityClasses() {
    List<Class<?>> entities = new ArrayList<>();
    entities.add(org.example.domain.Address.class);
    entities.add(org.example.domain.Contact.class);
    entities.add(org.example.domain.Content.class);
    entities.add(org.example.domain.Country.class);
    entities.add(org.example.domain.Customer.class);
    entities.add(org.example.domain.DMachine.class);
    entities.add(org.example.domain.DMachineStats.class);
    entities.add(org.example.domain.DMachineStatsAgg.class);
    entities.add(org.example.domain.Order.class);
    entities.add(org.example.domain.OrderDetail.class);
    entities.add(org.example.domain.OrderShipment.class);
    entities.add(org.example.domain.OrderTotals.class);
    entities.add(org.example.domain.Product.class);
    return entities;
  }

  @Override
  public List<Class<?>> entityClassesFor(String dbName) {

    return Collections.emptyList();
  }

}

```

And note that the related service loader file is generated into `META-INF/services/io.ebean.config.ModuleInfoLoader` which we can find in target/classes  (it isn't in target/generated-sources)

----
## Other (non-default) Databases
For entities that should register with a non-default database we expect those to have the `@DbName` annotation to specify the database name they relate to (and should be registered to).

e.g.
```java
@DbName("cust")
@Entity
@Table(name="customer")
public class Customer extends BaseModel {
```

When we use `@DbName` on Customer and Address entities then we see the generated code like:

```java
package org.example.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.annotation.Generated;

import io.ebean.config.ModuleInfo;
import io.ebean.config.ModuleInfoLoader;

@Generated("io.ebean.querybean.generator")
@ModuleInfo(entities={"cust:org.example.domain.Address","cust:org.example.domain.Customer","org.example.domain.Contact","org.example.domain.Content","org.example.domain.Country","org.example.domain.DMachine","org.example.domain.DMachineStats","org.example.domain.DMachineStatsAgg","org.example.domain.Order","org.example.domain.OrderDetail","org.example.domain.OrderShipment","org.example.domain.OrderTotals","org.example.domain.Product"})
public class _ebean$ModuleInfo implements ModuleInfoLoader {

  @Override
  public List<Class<?>> entityClasses() {
    List<Class<?>> entities = new ArrayList<>();
    entities.add(org.example.domain.Contact.class);
    entities.add(org.example.domain.Content.class);
    entities.add(org.example.domain.Country.class);
    entities.add(org.example.domain.DMachine.class);
    entities.add(org.example.domain.DMachineStats.class);
    entities.add(org.example.domain.DMachineStatsAgg.class);
    entities.add(org.example.domain.Order.class);
    entities.add(org.example.domain.OrderDetail.class);
    entities.add(org.example.domain.OrderShipment.class);
    entities.add(org.example.domain.OrderTotals.class);
    entities.add(org.example.domain.Product.class);
    return entities;
  }

  @Override
  public List<Class<?>> entityClassesFor(String dbName) {

    if ("cust".equals(dbName)) return cust_entities();
    return Collections.emptyList();
  }

  private List<Class<?>> cust_entities() {
    List<Class<?>> entities = new ArrayList<>();
    entities.add(org.example.domain.Address.class);
    entities.add(org.example.domain.Customer.class);
    return entities;
  }

}
```

-----
Note that as this is off by default initially we can put this into the next patch release `12.1.8`

**Comment by jnehlmeier** [comment]
Hmm what if the `@DbName` annotation is only used on `@MappedSuperClass` classes to avoid repetition? E.g. having a base model for database 1 and a base model for database 2 and then let entities subclass accordingly. That is what I usually do, as it saves me from annotating every entity.

Would the query bean annotation processor see sub classes then?

**Comment by rbygrave** [comment]
> Would the query bean annotation processor see sub classes then?

Yes it does that already, we can put `@DbName` on a common `@MappedSuperclass` and that is the likely / expected way to use it.

**Comment by jnehlmeier** [comment]
Just looked in our code, since I couldn't remember yesterday.

Looks like we haven't used `@DbName` at all. We use the `Model(String dbName)` super constructor call instead in our base models for the different databases. So maybe you should document the implications when using this constructor as it blocks you from using this new feature here.

**Comment by rbygrave** [comment]
> So maybe you should document the implications when using this constructor as it blocks you from using this new feature

No it should not.  I have not been clear.

Our entities extending Model ... still need to use that constructor `Model(String dbName)` when they relate to non-default database, there is no change there.

The `@DbName` was introduced such that the _generated query beans_ for non-default databases had a default constructor that uses the correct database.  Before this we always needed to new up those query beans using the constructor that took the Database (which was less than elegant).  

Using the `@DbName` didn't really do anything wrt the entity beans in that they still need to use `Model(String dbName)` - they still need to do that, no change there, you should still all be good.

-----
How using `@DbName` impacts entity beans is that we can (we will) generate code that we can optionally use to register them with (to avoid classpath scanning or maintaining that code ourselves).

> because the Finder class also has a constructor with a database name

Yes, and that is good and we need to continue doing that as I see it.  What using `@DbName` means is that again when we new up a query bean (for non default db) ... we can now use the default constructor.

-----
In summary for entities that "belong" to a non-default database.
- The entity beans ultimately use the `Model(String dbName)` constructor
- The related Finder (if exists) should similarly use the  `Finder(String dbName)` constructor
- Additionally we should put  `@DbName` on the entity beans (usually a mapped super) ... such that the generated query beans are nicer to use and (optional but nice) such that we can use generated code (ModuleInfoLoader) to register them 

It would be recommended to do all 3 of those things (for these entities belonging to non-default databases).

**Comment by jnehlmeier** [comment]
Ahh ok got it. Thought `@DbName` is not exclusive to query beans but would also be used by ebean itself. So I will add that annotation to our base models then and simplify our finders / query bean usages a bit. 👍

**Comment by richdrich** [comment]
As of `12.2.1` this is defaulted `true` and it breaks the loading of `@Embeddable` and `@Converter` modules because *only* modules from `ModuleInfoLoader` get loaded. I will dig further and may raise a bug.

**Comment by jnehlmeier** [comment]
@richdrich If you have multiple ebean databases make sure that your embeddable and possibly converter classes are annotated with `@DbName("....")`. Without this annotation the class will be assigned to the default database configured in ebean. This might be the wrong database or you might not have a default database configured at all (e.g. when having a multi tenant app with multiple databases)
